framework/contrib/hit/main.cc

```diff
@@ -4,132 +4,332 @@
 #include <sstream>
 #include <fstream>
 #include <map>
+#include <set>
 
 #include "parse.h"
 
-class MyWalker : public hit::Walker
+class Flags;
+std::vector<std::string> parseOpts(int argc, char ** argv, Flags & flags);
+
+int renameParam(int argc, char ** argv);
+int findParam(int argc, char ** argv);
+int validate(int argc, char ** argv);
+int format(int argc, char ** argv);
+
+int
+main(int argc, char ** argv)
+{
+  if (argc < 2)
+  {
+    std::cerr << "must specify a subcommand\n";
+    return 1;
+  }
+
+  std::string subcmd(argv[1]);
+
+  if (subcmd == "rename")
+    return renameParam(argc - 2, argv + 2);
+  else if (subcmd == "find")
+    return findParam(argc - 2, argv + 2);
+  else if (subcmd == "validate")
+    return validate(argc - 2, argv + 2);
+  else if (subcmd == "format")
+    return format(argc - 2, argv + 2);
+
+  std::cerr << "unrecognized subcommand '" << subcmd << "'\n";
+  return 1;
+}
+
+struct Flag
+{
+  bool arg;
+  bool have;
+  std::string val;
+  std::string help;
+};
+
+class Flags
 {
 public:
-  void walk(const std::string & fullpath, const std::string & nodepath, hit::Node * f)
+  void add(std::string name, std::string help, std::string def = "__NONE__")
   {
-    std::cout << "    path=" << fullpath << "/" << nodepath << "\n";
+    if (def == "__NONE__")
+      flags[name] = {false, false, "", help};
+    else
+      flags[name] = {true, false, def, help};
   }
+
+  bool have(std::string flag) { return flags.count(flag) > 0 && flags[flag].have; }
+  std::string val(std::string flag) { return flags[flag].val; }
+
+  std::map<std::string, Flag> flags;
 };
 
-class DupSectionWalker : public hit::Walker
+std::vector<std::string>
+parseOpts(int argc, char ** argv, Flags & flags)
+{
+  int i = 0;
+  for (; i < argc; i++)
+  {
+    std::string arg = argv[i];
+    if (arg[0] != '-')
+      break;
+
+    std::string flagname = arg.substr(1);
+    if (flagname[0] == '-')
+      flagname = flagname.substr(1);
+
+    if (flags.flags.count(flagname) == 0)
+      throw std::runtime_error("unknown flag '" + arg);
```

Comment:
If your compiler throws exceptions, do you have all the proper try/catch logic in MOOSE? This is important because we need to quit a failing program the proper way (e.g. MPI_Abort() or mooseError()).

Reply:
This isn't in MOOSE - it is a completely separate binary tool.  MOOSE doesn't run this binary for anything either.  If we want any of this functionality in MOOSE - the CLI flag handling would not be a part of that.

Reply:
Right, I know this is for the stand alone tool. It just made me think about how it might be used in MOOSE and wanted you to think if we were correctly catching exceptions and terminating properly.